plugins/callback/mail.py

```diff
@@ -100,21 +101,25 @@ def mail(self, subject='Ansible error mail', body=None):
 
         smtp = smtplib.SMTP(self.smtphost, port=self.smtpport)
 
-        content = 'From: %s\n' % self.sender
-        content += 'To: %s\n' % self.to
+        sender_address = email.utils.parseaddr(self.sender)
+        to_addresses = email.utils.getaddresses([self.to])
+        cc_addresses = email.utils.getaddresses([self.cc])
+        bcc_addresses = email.utils.getaddresses([self.bcc])
+
+        content = 'From: %s\n' % email.utils.formataddr(sender_address)
```

Comment:
As pointed out in the original PR, `formataddr()` raises an exception on [international emails](https://en.wikipedia.org/wiki/International_email), since it only accepts ASCII email addresses. So this change will actually break any user who relied on international emails working in the past.

I'm not really sure how to properly fix this, I know way too little about Python's `email` module.

Reply:
I think that international emails are requiring the use of the `SMTPUTF8` extension.
Also the Python’s `smtplib` module supports it, it requires to have it [explicitly specified in the `mail_options` parameter of the `sendmail` function](https://docs.python.org/3/library/smtplib.html#smtplib.SMTP.sendmail).
As we’re not passing `mail_options=["SMTPUTF8"]` to `sendmail`, I’m not sure international emails can work with the current version of the code.

Reply:
I did try it out (with `smtplib`), if I don't specify `mail_options=["SMTPUTF8"]`, the result is
```
[...]
  File "/usr/lib/python3.10/smtplib.py", line 892, in sendmail
    (code, resp) = self.rcpt(each, rcpt_options)
  File "/usr/lib/python3.10/smtplib.py", line 553, in rcpt
    self.putcmd("rcpt", "TO:%s%s" % (quoteaddr(recip), optionlist))
  File "/usr/lib/python3.10/smtplib.py", line 378, in putcmd
    self.send(f'{s}{CRLF}')
  File "/usr/lib/python3.10/smtplib.py", line 357, in send
    s = s.encode(self.command_encoding)
UnicodeEncodeError: 'ascii' codec can't encode character '\xeb' in position 10: ordinal not in range(128)
```
If I pass `mail_options=[...]`, it talks to the SMTP server (and I receive `smtplib.SMTPRecipientsRefused` since that email address doesn't exist on my server :D ).

So good, this problem is resolved :)

Reply:
Except that we should catch the exception that `formataddr` throws and throw a nicer `AnsibleCallbackError` instead.

Reply:
In your test case, the non-ascii character has been caught by the Python code.
But even without the Python-side check, the non-ascii character in the address would have been rejected by the SMTP server.

Without the `SMTPUTF8` option, the SMTP server issues a `Syntax error`:
```console
$ netcat gmail-smtp-in.l.google.com 25
220 mx.google.com ESMTP i12si2342221wrf.680 - gsmtp
EHLO me
250-mx.google.com at your service, [2a01:e0a:465:5440:9919:183e:d32a:e458]
250-SIZE 157286400
250-8BITMIME
250-STARTTLS
250-ENHANCEDSTATUSCODES
250-PIPELINING
250-CHUNKING
250 SMTPUTF8
MAIL FROM:<test@example.org>
250 2.1.0 OK i12si2342221wrf.680 - gsmtp
RCPT TO:<foo@Lénaïc.fr>
555 5.5.2 Syntax error. i12si2342221wrf.680 - gsmtp
QUIT
221 2.0.0 closing connection i12si2342221wrf.680 - gsmtp
```

Whereas, with the `SMTPUTF8` option, the SMTP server accepts to check the recipient address:
```console
$ netcat gmail-smtp-in.l.google.com 25
220 mx.google.com ESMTP n7si1985190wri.117 - gsmtp
EHLO me
250-mx.google.com at your service, [2a01:e0a:465:5440:9919:183e:d32a:e458]
250-SIZE 157286400
250-8BITMIME
250-STARTTLS
250-ENHANCEDSTATUSCODES
250-PIPELINING
250-CHUNKING
250 SMTPUTF8
MAIL FROM:<test@example.org> SMTPUTF8
250 2.1.0 OK n7si1985190wri.117 - gsmtp
RCPT TO:<foo@Lénaïc.fr>
550-5.1.1 The email account that you tried to reach does not exist. Please try
550-5.1.1 double-checking the recipient's email address for typos or
550-5.1.1 unnecessary spaces. Learn more at
550 5.1.1  https://support.google.com/mail/?p=NoSuchUser n7si1985190wri.117 - gsmtp
QUIT
221 2.0.0 closing connection n7si1985190wri.117 - gsmtp
```

In odrer to add support of IDN, we could systematically encode the domain part of the email addresses with `str.encode("idna")`.
It would look something like:
```python
def _formataddr(pair):
    realname, address = pair
    user, domain = address.split('@')
    domain = domain.encode('idna').decode('ascii')
    return email.utils.formataddr((realname, '%s@%s' % (user, domain)))

addresses = email.utils.getaddresses([original_string])

encoded_addresses = ', '.join([_formataddr(pair) for pair in addresses])
```

It would add support of IDN without requiring the use of the `SMTPUTF8` option.
It would however not help for the `user` part of the email.